File: svuaback/src/main/java/cl/aracridav/svua/depreciacion/service/DepreciacionService.java
package cl.aracridav.svua.depreciacion.service;

import java.util.List;

import cl.aracridav.svua.depreciacion.entity.DepreciacionMensual;
import cl.aracridav.svua.depreciacion.entity.TipoDepreciacion;
import cl.aracridav.svua.inventario.activo.entity.Activo;

public interface DepreciacionService {

    public void guardarDepreciacion(Activo activo);

    public void calcularYGuardarDepreciacionMensual(Activo activo);

    public void calcularYGuardarDepreciacionMensual(Activo activo, Long empresaId);

    public List<DepreciacionMensual> obtenerDepreciacionesPorActivo(Activo activo);

    public List<DepreciacionMensual> obtenerDepreciacionesPorActivo(Activo activo, TipoDepreciacion tipo);

    // Backfill: genera el cronograma ACELERADA para los activos de la
    // empresa actual que aún no lo tienen (creados antes de que
    // existiera, o cargados por importación masiva). Idempotente.
    // Devuelve la cantidad de activos procesados.
    public int generarDepreciacionAceleradaFaltante();

    // Genera el cronograma ACELERADA para un activo puntual. Falla si
    // el activo ya lo tiene calculado (evita duplicarlo).
    public void generarDepreciacionAceleradaPorActivo(Long activoId);

}

File: frontend/src/app/services/depreciacion.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

// 🔥 Backfill de depreciación acelerada tributaria (SII) para activos
// que quedaron sin ese cronograma (creados antes de que existiera, o
// cargados por Excel). Ver DepreciacionController en el backend.
@Injectable({
  providedIn: 'root'
})
export class DepreciacionService {

  private apiUrl = environment.apiUrl;
  private http = inject(HttpClient);

  // Genera la depreciación acelerada para todos los activos de la
  // empresa actual que aún no la tienen. Devuelve la cantidad procesada.
  generarAceleradaFaltante(): Observable<number> {
    return this.http.post<number>(
      `${this.apiUrl}/depreciacion/acelerada/generar-faltantes`,
      null
    );
  }

  // Genera la depreciación acelerada de un activo puntual. El backend
  // rechaza la llamada (400) si el activo ya la tenía calculada.
  generarAceleradaPorActivo(activoId: number): Observable<void> {
    return this.http.post<void>(
      `${this.apiUrl}/depreciacion/acelerada/generar/${activoId}`,
      null
    );
  }
}
